Feature/normalize input

PReichetanz opened the pull request:
Add functionality to normalize name input before saving to localStorage

- whitespace will be removed
- first letter uppercase, all other letters lowercase
- name is saved as first, middle, and last name
- Update Class- and PupilOverview to show the name accordingly

[View at heroku](https://capstone-rev-feature-no-q8ytkq.herokuapp.com/)

PReichetanz committed:
Add feature: normalize name input

- Add normalizeInput.ts and functionality
- Update name type
- Add types: RawInput, NormalizedInput, PupilName

PReichetanz committed:
Update Card, Class- and PupilOverview

- Add new types
- Add display of first, middle, and last name
- Change Card key prop to pupil.id

PReichetanz committed:
Update Card stories

PReichetanz committed:
Rename const whitespaceInput to replacedHyphens

F-Kirchhoff commented on src/app/components/Card/Card.tsx:
```diff
@@ -18,7 +18,7 @@ export default function Card({ pupil, deleteCard }: CardProps): JSX.Element {
       />
       <CardLink to={`/pupil/${pupil.id}`}>
-        <Heading>{name}</Heading>
+        <Heading>{`${name.first} ${name.middle} ${name.last}`}</Heading>
```
I like the naming of the name props

F-Kirchhoff commented on src/app/components/utils/normalizeInput.ts:
```diff
@@ -0,0 +1,69 @@
+}
+
+function prepareNormalisation(name: string) {
+  return name.trim().replace(/\s+/g, ' ').split(' ');
```
maybe you can give us a short workshop on regex? :D

F-Kirchhoff commented on src/app/components/utils/normalizeInput.ts:
```diff
@@ -0,0 +1,69 @@
+  const preparedName = prepareNormalisation(pupil.name);
+  const firstName = preparedName[0];
+  const middleName =
+    preparedName.length < 2 ? '' : preparedName.slice(1, -1).join(' ');
```
maybe for a more semantic code you could set the middle name to none if its not there

F-Kirchhoff commented on src/app/components/utils/normalizeInput.ts:
```diff
@@ -0,0 +1,69 @@
+    preparedName.length < 2 ? '' : preparedName.slice(1, -1).join(' ');
+  const lastName =
+    preparedName.length === 1
+      ? ''
```
same here

F-Kirchhoff commented on src/app/components/utils/normalizeInput.ts:
```diff
@@ -0,0 +1,69 @@
+      ? ''
+      : preparedName[preparedName.length - 1].toLowerCase();
+
+  const category = pupil.category;
+  const evaluation = pupil.evaluation;
```
you can use the destructuring here: const {category, evaluation } = pupil

F-Kirchhoff commented on src/app/components/utils/normalizeInput.ts:
```diff
@@ -0,0 +1,69 @@
+    inputToNormalize = replacedWhitespace;
+    return inputToNormalize;
+  }
+  const normalizedInput = toUpperCaseWhitespace(inputToNormalize, 0);
+  return normalizedInput;
```
just use: return toUpperCaseWhitespace(inputToNormalize, 0);

F-Kirchhoff reviewed (approved):
lgtm :) and really complex too. Had some ideas for further improvement.

derjubu reviewed (approved):
LGTM 🚀

PReichetanz committed:
Changes according to review

- Set value of middle and last name to null if nothing is submitted
- Desctructure category and evaluation from pupil
- Refactor return statement (delete const declaration)
